// xstate/lib/test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var graph_1 = require("./graph");
var Machine_1 = require("./Machine");
var _1 = require(".");
var utils_1 = require("./utils");
var adjMapCache = new WeakMap();
var defaultReachesOptions = {
    filter: function () { return true; }
};
function reaches(machine, stateValue, options) {
    if (options === void 0) { options = defaultReachesOptions; }
    console.dir(graph_1.getSimplePaths(machine), { depth: null });
    if (!adjMapCache.get(machine)) {
        adjMapCache.set(machine, graph_1.getSimplePaths(machine));
    }
    var adjMap = adjMapCache.get(machine);
    var matchingKeys = Object.keys(adjMap).filter(function (key) {
        var _a = graph_1.deserializeStateString(key), value = _a.value, context = _a.context;
        return (utils_1.matchesState(stateValue, value) &&
            (!options.filter || options.filter({ value: value, context: context })));
    });
    console.log(matchingKeys);
    return !!matchingKeys.length;
}
exports.reaches = reaches;
var testMachine = Machine_1.Machine({
    id: 'light',
    initial: 'green',
    context: { count: 10 },
    states: {
        green: {
            on: {
                TIMER: {
                    target: 'yellow.whatever',
                    actions: _1.assign({ count: 25 })
                }
            }
        },
        yellow: {
            initial: 'whatever',
            states: {
                whatever: {
                    initial: 'deep',
                    states: {
                        deep: {}
                    }
                },
                yeah: {}
            }
        }
    }
});
var r = reaches(testMachine, { yellow: 'whatever' }, {
    filter: function (_a) {
        var context = _a.context;
        return context.count > 20;
    }
});
console.log(r);
