// xstate/lib/test/examples/6.6.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../../src/index");
var utils_1 = require("../utils");
describe('Example 6.6', function () {
    var machine = index_1.Machine({
        initial: 'A',
        states: {
            A: {
                on: {
                    3: 'B'
                },
                initial: 'D',
                states: {
                    C: {
                        on: {
                            2: '#B'
                        }
                    },
                    D: {
                        on: {
                            1: 'C'
                        }
                    }
                }
            },
            B: {
                id: 'B',
                on: {
                    4: 'A.D'
                }
            }
        }
    });
    var expected = {
        A: {
            1: 'A.C',
            2: 'A.D',
            3: 'B',
            4: 'A.D'
        },
        B: {
            1: 'B',
            2: 'B',
            3: 'B',
            4: 'A.D'
        },
        'A.C': {
            1: 'A.C',
            2: 'B',
            3: 'B',
            4: 'A.C'
        },
        'A.D': {
            1: 'A.C',
            2: 'A.D',
            3: 'B',
            4: 'A.D'
        }
    };
    utils_1.testAll(machine, expected);
});
