// xstate/lib/test/examples/6.8.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var index_1 = require("../../src/index");
var utils_1 = require("../utils");
describe('Example 6.8', function () {
    var machine = index_1.Machine({
        initial: 'A',
        states: {
            A: {
                on: {
                    6: 'F'
                },
                initial: 'B',
                states: {
                    B: {
                        on: { 1: 'C' }
                    },
                    C: {
                        on: { 2: 'E' }
                    },
                    D: {
                        on: { 3: 'B' }
                    },
                    E: {
                        on: { 4: 'B', 5: 'D' }
                    },
                    hist: { history: true }
                }
            },
            F: {
                on: {
                    5: 'A.hist'
                }
            }
        }
    });
    var expected = {
        A: {
            1: 'A.C',
            6: 'F'
        },
        'A.B': {
            1: 'A.C',
            6: 'F',
            FAKE: undefined
        },
        'A.C': {
            2: 'A.E',
            6: 'F',
            FAKE: undefined
        },
        'A.D': {
            3: 'A.B',
            6: 'F',
            FAKE: undefined
        },
        'A.E': {
            4: 'A.B',
            5: 'A.D',
            6: 'F',
            FAKE: undefined
        },
        F: {
            5: 'A.B'
        }
    };
    utils_1.testAll(machine, expected);
    it('should respect the history mechanism', function () {
        var stateC = machine.transition('A.B', '1');
        var stateF = machine.transition(stateC, '6');
        var stateActual = machine.transition(stateF, '5');
        chai_1.assert.deepEqual(stateActual.value, { A: 'C' });
    });
});
