// xstate/lib/packages/xstate-graph/src/graph.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var xstate_1 = require("xstate");
var utils_1 = require("xstate/lib/utils");
var actions_1 = require("xstate/lib/actions");
var EMPTY_MAP = {};
function getNodes(node) {
    var states = node.states;
    var nodes = utils_1.keys(states).reduce(function (accNodes, stateKey) {
        var subState = states[stateKey];
        var subNodes = getNodes(states[stateKey]);
        accNodes.push.apply(accNodes, __spread([subState], subNodes));
        return accNodes;
    }, []);
    return nodes;
}
exports.getNodes = getNodes;
function getEventEdges(node, event) {
    var transitions = node.definition.on[event];
    return utils_1.flatten(transitions.map(function (transition) {
        var targets = transition.target
            ? [].concat(transition.target)
            : undefined;
        if (!targets) {
            return [
                {
                    source: node,
                    target: node,
                    event: event,
                    actions: transition.actions
                        ? transition.actions.map(utils_1.getActionType)
                        : [],
                    cond: transition.cond,
                    transition: transition
                }
            ];
        }
        return targets
            .map(function (target) {
            try {
                var targetNode = target
                    ? node.getRelativeStateNodes(target, undefined, false)[0]
                    : node;
                return {
                    source: node,
                    target: targetNode,
                    event: event,
                    actions: transition.actions
                        ? transition.actions.map(utils_1.getActionType)
                        : [],
                    cond: transition.cond,
                    transition: transition
                };
            }
            catch (e) {
                // tslint:disable-next-line:no-console
                console.warn("Target '" + target + "' not found on '" + node.id + "'");
                return undefined;
            }
        })
            .filter(function (maybeEdge) { return maybeEdge !== undefined; });
    }));
}
exports.getEventEdges = getEventEdges;
function getEdges(node, options) {
    var _a = (options || {}).depth, depth = _a === void 0 ? null : _a;
    var edges = [];
    if (node.states && depth === null) {
        utils_1.keys(node.states).forEach(function (stateKey) {
            edges.push.apply(edges, __spread(getEdges(node.states[stateKey])));
        });
    }
    else if (depth && depth > 0) {
        utils_1.keys(node.states).forEach(function (stateKey) {
            edges.push.apply(edges, __spread(getEdges(node.states[stateKey], { depth: depth - 1 })));
        });
    }
    utils_1.keys(node.on).forEach(function (event) {
        edges.push.apply(edges, __spread(getEventEdges(node, event)));
    });
    return edges;
}
exports.getEdges = getEdges;
function getAdjacencyMap(node, context) {
    var adjacency = {};
    var events = node.events;
    function findAdjacencies(stateValue) {
        var e_1, _a;
        var stateKey = JSON.stringify(stateValue);
        if (adjacency[stateKey]) {
            return;
        }
        adjacency[stateKey] = {};
        try {
            for (var events_1 = __values(events), events_1_1 = events_1.next(); !events_1_1.done; events_1_1 = events_1.next()) {
                var event_1 = events_1_1.value;
                var nextState = node.transition(stateValue, event_1, context);
                adjacency[stateKey][event_1] = { state: nextState.value };
                findAdjacencies(nextState.value);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (events_1_1 && !events_1_1.done && (_a = events_1.return)) _a.call(events_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    findAdjacencies(node.initialState.value);
    return adjacency;
}
exports.getAdjacencyMap = getAdjacencyMap;
function deserializeStateString(valueContextString) {
    var _a = __read(valueContextString.split(' | '), 2), valueString = _a[0], contextString = _a[1];
    return {
        value: JSON.parse(valueString),
        context: contextString === undefined ? undefined : JSON.parse(contextString)
    };
}
exports.deserializeStateString = deserializeStateString;
function serializeState(state) {
    var value = state.value, context = state.context;
    return context === undefined
        ? JSON.stringify(value)
        : JSON.stringify(value) + ' | ' + JSON.stringify(context);
}
exports.serializeState = serializeState;
function serializeEvent(event) {
    return JSON.stringify(event);
}
exports.serializeEvent = serializeEvent;
function deserializeEventString(eventString) {
    return JSON.parse(eventString);
}
exports.deserializeEventString = deserializeEventString;
var defaultValueAdjMapOptions = {
    events: {},
    filter: function () { return true; },
    stateSerializer: serializeState,
    eventSerializer: serializeEvent
};
var ValueAdjacency = /** @class */ (function () {
    function ValueAdjacency(machine, options) {
        this.machine = machine;
        this.options = __assign({ events: {}, stateSerializer: serializeState, eventSerializer: serializeEvent }, options);
        this.mapping = getValueAdjacencyMap(machine, options);
    }
    ValueAdjacency.prototype.reaches = function (stateValue, context) {
        var resolvedStateValue = this.machine.resolve(stateValue);
        var state = xstate_1.State.from(resolvedStateValue, context);
        return !!this.mapping[this.options.stateSerializer(state)];
    };
    return ValueAdjacency;
}());
exports.ValueAdjacency = ValueAdjacency;
function getValueAdjacencyMap(node, options) {
    var optionsWithDefaults = __assign({ events: {}, stateSerializer: serializeState, eventSerializer: serializeEvent }, options);
    var filter = optionsWithDefaults.filter, stateSerializer = optionsWithDefaults.stateSerializer, eventSerializer = optionsWithDefaults.eventSerializer;
    var events = {};
    node.events.forEach(function (event) {
        events[event] = [event];
    });
    Object.assign(events, optionsWithDefaults.events);
    var adjacency = {};
    function findAdjacencies(state) {
        var e_2, _a;
        var nextEvents = state.nextEvents;
        var stateHash = stateSerializer(state);
        if (adjacency[stateHash]) {
            return;
        }
        adjacency[stateHash] = {};
        var potentialEvents = utils_1.flatten(nextEvents.map(function (nextEvent) { return events[nextEvent] || []; })).map(function (event) { return actions_1.toEventObject(event); });
        try {
            for (var potentialEvents_1 = __values(potentialEvents), potentialEvents_1_1 = potentialEvents_1.next(); !potentialEvents_1_1.done; potentialEvents_1_1 = potentialEvents_1.next()) {
                var event_2 = potentialEvents_1_1.value;
                var nextState = node.transition(state, event_2);
                if ((!filter || filter(nextState)) &&
                    stateHash !== stateSerializer(nextState)) {
                    adjacency[stateHash][eventSerializer(event_2)] = nextState;
                    findAdjacencies(nextState);
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (potentialEvents_1_1 && !potentialEvents_1_1.done && (_a = potentialEvents_1.return)) _a.call(potentialEvents_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    }
    findAdjacencies(node.initialState);
    return adjacency;
}
exports.getValueAdjacencyMap = getValueAdjacencyMap;
function getShortestValuePaths(machine, options) {
    if (options === void 0) { options = defaultValueAdjMapOptions; }
    if (!machine.states) {
        return EMPTY_MAP;
    }
    var adjacency = getValueAdjacencyMap(machine, options);
    var pathMap = {};
    var visited = new Set();
    function util(state) {
        var e_3, _a, e_4, _b;
        var stateKey = serializeState(state);
        visited.add(stateKey);
        var eventMap = adjacency[stateKey];
        try {
            for (var _c = __values(utils_1.keys(eventMap)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var eventType = _d.value;
                var _e = eventMap[eventType], value = _e.value, context_1 = _e.context;
                if (!value) {
                    continue;
                }
                var nextState = xstate_1.State.from(value, context_1);
                var nextStateId = serializeState(nextState);
                if (!pathMap[nextStateId] ||
                    pathMap[nextStateId].length > pathMap[stateKey].length + 1) {
                    pathMap[nextStateId] = __spread((pathMap[stateKey] || []), [
                        {
                            state: { value: value, context: state.context },
                            event: deserializeEventString(eventType)
                        }
                    ]);
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_3) throw e_3.error; }
        }
        try {
            for (var _f = __values(utils_1.keys(eventMap)), _g = _f.next(); !_g.done; _g = _f.next()) {
                var event_3 = _g.value;
                var _h = eventMap[event_3], value = _h.value, context_2 = _h.context;
                if (!value) {
                    continue;
                }
                var nextState = xstate_1.State.from(value, context_2);
                var nextStateId = serializeState(xstate_1.State.from(value, context_2));
                if (visited.has(nextStateId)) {
                    continue;
                }
                util(nextState);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return pathMap;
    }
    util(machine.initialState);
    return pathMap;
}
exports.getShortestValuePaths = getShortestValuePaths;
function getShortestPathsAsArray(machine) {
    var result = getShortestValuePaths(machine, defaultValueAdjMapOptions);
    return utils_1.keys(result).map(function (key) { return ({
        state: JSON.parse(key),
        path: result[key]
    }); });
}
exports.getShortestPathsAsArray = getShortestPathsAsArray;
function getSimplePaths(machine, options) {
    if (!machine.states) {
        return EMPTY_MAP;
    }
    var adjacency = getValueAdjacencyMap(machine, options);
    var visited = new Set();
    var path = [];
    var paths = {};
    function util(fromStateSerial, toStateSerial) {
        var e_5, _a;
        visited.add(fromStateSerial);
        if (fromStateSerial === toStateSerial) {
            if (!paths[toStateSerial]) {
                paths[toStateSerial] = {
                    state: deserializeStateString(toStateSerial),
                    paths: []
                };
            }
            paths[toStateSerial].paths.push(__spread(path));
        }
        else {
            try {
                for (var _b = __values(utils_1.keys(adjacency[fromStateSerial])), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var subEvent = _c.value;
                    var nextState = adjacency[fromStateSerial][subEvent];
                    if (!nextState) {
                        continue;
                    }
                    var nextStateSerial = serializeState(nextState);
                    if (!visited.has(nextStateSerial)) {
                        path.push({
                            state: deserializeStateString(fromStateSerial),
                            event: deserializeEventString(subEvent)
                        });
                        util(nextStateSerial, toStateSerial);
                    }
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        path.pop();
        visited.delete(fromStateSerial);
    }
    var initialStateSerial = serializeState(machine.initialState);
    utils_1.keys(adjacency).forEach(function (nextStateSerial) {
        util(initialStateSerial, nextStateSerial);
    });
    return paths;
}
exports.getSimplePaths = getSimplePaths;
function getSimplePathsAsArray(machine, options) {
    var result = getSimplePaths(machine, options);
    return utils_1.keys(result).map(function (key) { return result[key]; });
}
exports.getSimplePathsAsArray = getSimplePathsAsArray;
