// xstate/lib/test/types.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var index_1 = require("../src/index");
function noop(_x) {
    return;
}
describe('StateSchema', function () {
    var lightMachine = index_1.Machine({
        key: 'light',
        initial: 'green',
        meta: { interval: 1000 },
        context: { elapsed: 0 },
        states: {
            green: {
                meta: { name: 'greenLight' },
                on: {
                    TIMER: 'yellow',
                    POWER_OUTAGE: 'red'
                }
            },
            yellow: {
                on: {
                    TIMER: 'red',
                    POWER_OUTAGE: 'red'
                }
            },
            red: {
                on: {
                    TIMER: 'green',
                    POWER_OUTAGE: 'red'
                },
                initial: 'walk',
                states: {
                    walk: {
                        on: {
                            PED_COUNTDOWN: 'wait'
                        }
                    },
                    wait: {
                        on: {
                            PED_COUNTDOWN: {
                                target: 'stop',
                                cond: function (ctx, e) {
                                    return e.duration === 0 && ctx.elapsed > 0;
                                }
                            }
                        }
                    },
                    stop: {
                        on: {
                            '': { target: 'green' }
                        }
                    }
                }
            }
        }
    });
    noop(lightMachine);
    it('should work with a StateSchema defined', function () {
        chai_1.assert.ok(true, 'Tests will not compile if types are wrong');
    });
});
